### PR TITLE
Improve ghe support

### PR DESCRIPTION
**Pull request by yashigani**
# Problem
GitHub Issue Badges for GHE support only one ghe ( #7 ) .

# Change
- Improve options page
  - Add UI to manage multiple GHE

## screen shot(GHE)
image

### CODE REVIEW

File: src/ts/app.tsx [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+function isConfig(config: any): config is Config {
+  return config.origin !== undefined && config.apiRoot !== undefined && config.token !== undefined;
+}
+
+const origins: any[] = JSON.parse(localStorage.getItem('origins') || '[]');
+const configs: Config[] = origins.filter(isConfig);
+const github: boolean = ((mode?: string) => {
+  if (mode) {
+    return mode == 'github';
```
Comment (grassedge) [inline src/ts/app.tsx]: `===` is better than `==` .

File: src/ts/app.tsx [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+  }
+  return false;
+})(localStorage.getItem('mode'));
```
Comment (grassedge) [inline src/ts/app.tsx]: I suppose that following code is enough.

```javascript
const github = (localStorage.getItem('mode') === 'github');
```

File: src/ts/app.tsx [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    };
+
+    this.addItem = this.addItem.bind(this);
+    this.removeItem = this.removeItem.bind(this);
+    this.updateStatus = this.updateStatus.bind(this);
+    this.saveConfigs = this.saveConfigs.bind(this);
+    this.onConfigChanged = this.onConfigChanged.bind(this);
+  }
+
+  addItem() {
```
Comment (grassedge) [inline src/ts/app.tsx]:
```javascript
class App {
  constructor() {
    this.addItem = this.addItem.bind(this);
    ...
  }
  addItem() { ... }
}
```

Above code can be converted to 

```javascript
class App {
  addItem = () => { ... }
}
```

typescript compiler outputs following code.

```javascript
function App() {
  this.addItem = () => { ... };
  ...
}
```

so both above code behave the same action.
removeItem/updateStatus/saveConfigs/onConfigChanged also.

File: src/ts/app.tsx [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    const newGithub = this.state.newGithub;
```
Comment (grassedge) [inline src/ts/app.tsx]: [may] if `newGithub` is false, you can return immediatly.

File: src/ts/app.tsx [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    const apiRoot = (newGithub === 'https://github.com') ? 'https://api.github.com' : (newGithub + '/api/v3')
+    if (newGithub !== '' && this.state.configs.filter(({origin}) => (origin === newGithub)).isEmpty) {
+      let config: Config = { origin: newGithub, apiRoot: apiRoot, token: '', status: TokenStatus.unchecked }
```
Comment (grassedge) [inline src/ts/app.tsx]: `config` is imutable, so you can use `const` instead of `let`.

File: src/ts/app.tsx [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    }
+  }
+
+  removeItem(i: number) {
+    this.setState(({configs}) => ({
+      configs: configs.delete(i)
+    }));
+  }
+
+  updateStatus(i: number) {
```
Comment (grassedge) [inline src/ts/app.tsx]: I think `updateTokenStatus` is better, because `status` and `this.state` is confusingly similar.

File: src/ts/app.tsx [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+      const newConfig = Map(config)
+        .update('status', s => status)
+        .toJS();
+      this.setState(({configs}) => ({
+        configs: configs.set(i, newConfig),
+      }));
+    };
+
+    const config = this.state.configs.get(i);
+    if (config === undefined && config.apiRoot === undefined) {
```
Comment (grassedge) [inline src/ts/app.tsx]: not `&&` but `||` ?

File: src/ts/app.tsx [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    if (config.token === '') {
+      updateStatus(config, TokenStatus.unchecked);
+      alert('please input your access token');
+      return;
+    }
+
+    const xhr = new XMLHttpRequest();
+    xhr.open("GET", config.apiRoot);
+    xhr.setRequestHeader("Authorization", `token ${config.token}`)
+    xhr.onload = function(e) {
```
Comment (grassedge) [inline src/ts/app.tsx:90]: Displaying `xhr.responseText` is helpful for users to understand what to do next.

**Review by grassedge — CHANGES_REQUESTED** [review]
Thanks for your contribution!
I commented on your code, please check it.

**Comment by yashigani** [comment]
@hatz48 Thanks your review. I tried to change the code.

**Comment by grassedge** [comment]
Looks good to me !

**Review by grassedge — APPROVED** [review]

**Comment by grassedge** [comment]
thx!

**Comment by motemen** [comment]
Thanks, just uploaded to Chrome store as v2.5.0.

**Pull request merged**